google/auth/transport/aiohttp_req.py

```diff
@@ -0,0 +1,337 @@
+            six.raise_from(new_exc, caught_exc)
+        """
+
+    async def request(
+        self,
+        method,
+        url,
+        data=None,
+        headers=None,
+        max_allowed_time=None,
+        timeout=_DEFAULT_TIMEOUT,
+        **kwargs
+    ):
+
+        # Use a kwarg for this instead of an attribute to maintain
+        # thread-safety.
+        _credential_refresh_attempt = kwargs.pop("_credential_refresh_attempt", 0)
+        # Make a copy of the headers. They will be modified by the credentials
+        # and we want to pass the original headers if we recurse.
+        request_headers = headers.copy() if headers is not None else {}
+
+        # Do not apply the timeout unconditionally in order to not override the
+        # _auth_request's default timeout.
+        auth_request = (
+            self._auth_request
+            if timeout is None
+            else functools.partial(self._auth_request, timeout=timeout)
+        )
+
+        remaining_time = max_allowed_time
+
+        with TimeoutGuard(remaining_time) as guard:
+            self.credentials.before_request(auth_request, method, url, request_headers)
+
+        with TimeoutGuard(remaining_time) as guard:
+            response = await super(AuthorizedSession, self).request(
+                method,
+                url,
+                data=data,
+                headers=request_headers,
+                timeout=timeout,
+                **kwargs
+            )
+
+        remaining_time = guard.remaining_timeout
+
+        if (
+            response.status in self._refresh_status_codes
+            and _credential_refresh_attempt < self._max_refresh_attempts
+        ):
+
+            _LOGGER.info(
+                "Refreshing credentials due to a %s response. Attempt %s/%s.",
+                response.status,
+                _credential_refresh_attempt + 1,
+                self._max_refresh_attempts,
+            )
+
+            # Do not apply the timeout unconditionally in order to not override the
+            # _auth_request's default timeout.
+            auth_request = (
+                self._auth_request
+                if timeout is None
+                else functools.partial(self._auth_request, timeout=timeout)
+            )
+
+            with TimeoutGuard(remaining_time) as guard:
+                async with self._refresh_lock:
+                    await self._loop.run_in_executor(
+                        None, self.credentials.refresh, auth_request
+                    )
+
+            remaining_time = guard.remaining_timeout
+
+            return await self.request(
+                method,
+                url,
+                data=data,
+                headers=headers,
+                max_allowed_time=remaining_time,
+                timeout=timeout,
+                _credential_refresh_attempt=_credential_refresh_attempt + 1,
+                **kwargs
+            )
+
+        await self._auth_request_session.close()
```

Comment:
As in my comment about this in the test, I wonder if this needs a context manager or something similar to handle teardown.

Reply:
Yes, could we discuss this in more detail? Currently I just close the sessions, but it would be good to discuss with both both of you (@crwilcox) if this can be done in a better way as you mentioned in the earlier comment.

Reply:
I see that we are closing the session on the instance `self`. Is this an instance member on the sync version as well?